File: Python/SOLID/3. LiskovSubstitution.py
class Rectangle:
    def __init__(self, width, height):
        self._height = height
        self._width = width

    @property
    def width(self):
        return self._width

    @width.setter
    def width(self, value):
        self._width = value

    @property
    def height(self):
        return self._height

    @height.setter
    def height(self, value):
        self._height = value

    @property
    def area(self):
        return self._width * self._height

    def __str__(self):
        return f"Width: ${self._width}, height: ${self._height}"


class Square(Rectangle):
    def __init__(self, size):
        Rectangle.__init__(self, size, size)

    @Rectangle.width.setter
    def width(self, value):
        self._width = self._height = value

    @Rectangle.height.setter
    def height(self, value):
        self._width = self._height = value


def useIt(rc):
    w = rc.width
    rc.height = 10  # violation of LSP as the inherited square class does not behave the same way as rectangle!
    expected = int(w * 10)
    print(f"Expected area {expected}, but got {rc.area}")


rc = Rectangle(2, 3)
useIt(rc)

sq = Square(5)
useIt(sq)

File: Javascript/SOLID/3. LiskovSubstitution.js
class Rectangle {
  constructor(width, height) {
    this._width = width;
    this._height = height;
  }

  get width() {
    return this._width;
  }

  get height() {
    return this._height;
  }

  set width(value) {
    this._width = value;
  }

  set height(value) {
    this._height = value;
  }

  get area() {
    return this._width * this._height;
  }

  toString() {
    return `${this._width} x ${this._height}`;
  }
}

// Square class not required. It violates LSP
class Square extends Rectangle {
  constructor(size) {
    super(size, size);
  }

  set width(value) {
    this._width = this._height = value;
  }

  set height(value) {
    this._width = this._height = value;
  }
}

let useIt = function (rc) {
  let width = rc._width;
  rc.height = 10;
  let expected = width * 10;
  console.log(`Expected area ${expected}, but got ${rc.area}`);
};

let rc = new Rectangle(2, 3);
useIt(rc);

let sq = new Square(5);
useIt(sq);

// sq.width = 10;
// console.log(sq.toString());
